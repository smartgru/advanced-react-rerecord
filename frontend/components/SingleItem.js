import React, { Component } from 'react';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import styled from 'styled-components';
import Head from 'next/head';
import Error from './ErrorMessage';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    Item(where: { id: $id }) {
      id
      name
      description
      image {
        publicUrlTransformed
      }
    }
  }
`;

function SingleItem({ id }) {
  const { loading, error, data } = useQuery(SINGLE_ITEM_QUERY, {
    variables: { id },
  });
  if (error) return <Error error={error} />;
  if (loading) return <p>Loading...</p>;
  if (!data.Item) return <p>No Item Found for {id}</p>;
  const { Item } = data;
  return (
    <SingleItemStyles>
      <Head>
        <title>Sick Fits | {Item.name}</title>
      </Head>
      <img src={Item.image.publicUrlTransformed} alt={Item.name} />
      <div className="details">
        <h2>Viewing {Item.name}</h2>
        <p>{Item.description}</p>
      </div>
    </SingleItemStyles>
  );
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };
