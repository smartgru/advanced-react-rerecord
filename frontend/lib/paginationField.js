export default function paginationField() {
  return {
    keyArgs: false, // take full control of this field
    // We write custom functions to merge and read based on the 'first' and 'skip' args
    // first = 4 items per page
    // skip = how many items to offset
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      console.log('merging...', { first, skip, existing, incoming });

      const merged = existing ? existing.slice(0) : [];
      // we do it this way because someone might visit page 2 first, so we need to pad blank spots in the array
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
    read(existing = [], { args }) {
      // When we fire off a query, Apollo will check the cache first. We control that with a read method
      const { skip, first } = args;
      console.log('reading...', { args });
      console.log(existing);
      // See if we have the items we want
      const items = existing
        .slice(args.skip, args.skip + args.first)
        // we filter for empty spots because its likely we have padded spots with nothing in them.
        .filter(x => x);
      // If there are items, and they aren't blank spots, return them.
      if (items.length) {
        console.log('We have items! Gonna serve them from the cache', {
          items,
        });
        return items;
      }
      // Otherwise this function returns undefined and it will hit the network for the items, and call merge() for us
    },
  };
}
