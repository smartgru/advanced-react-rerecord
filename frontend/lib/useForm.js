import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    updateInputs({
      ...inputs,
      [e.target.name]: e.target.value,
    });
  }

  function resetForm() {
    // oh - just use the default state passed in
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm,
  };
}
