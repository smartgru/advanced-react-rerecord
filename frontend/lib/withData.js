import withApollo from 'next-with-apollo';
import { ApolloClient, InMemoryCache, ApolloLink } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { createUploadLink } from 'apollo-upload-client';
import { getDataFromTree } from '@apollo/react-ssr';
import { endpoint, prodEndpoint } from '../config';
import paginationField from './paginationField';

function createClient({ headers, initialState }) {
  console.log('initialState');
  console.log(initialState);
  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError) console.log(`[Network error]: ${networkError}`);
      }),
      // this uses apollo-link-http under the hood, so all the options here come from that package
      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allItems: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
