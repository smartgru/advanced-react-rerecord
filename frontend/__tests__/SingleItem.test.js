import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MockedProvider } from '@apollo/react-testing';
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { fakeItem } from '../lib/testUtils';

const item = fakeItem();
describe('<SingleItem/>', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // return this fake data (mocked data)
        result: {
          data: {
            Item: item,
          },
        },
      },
    ];
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    await screen.findByTestId('singleItem');
    expect(container).toMatchSnapshot();
  });

  it('Errors with a not found item', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Items Not Found!' }],
        },
      },
    ];
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );

    await screen.findByTestId('graphql-error');
    expect(container).toHaveTextContent('Items Not Found!');
  });
});
