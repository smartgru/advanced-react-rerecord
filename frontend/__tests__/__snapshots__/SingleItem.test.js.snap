// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> renders with proper data 1`] = `
<div>
  <div
    class="SingleItem__SingleItemStyles-mflgtv-0 AQPQe"
    data-testid="singleItem"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        Viewing 
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
